common/src/project.rs

```diff
@@ -76,35 +76,46 @@ impl AsRef<String> for ProjectName {
     }
 }
 
+impl<'de> Deserialize<'de> for ProjectName {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        let s: String = String::deserialize(deserializer)?;
+        s.parse().map_err(DeError::custom)
+    }
+}
```

Comment:
There is a problem with this, and that is that this will just return 400 and the message from the gateway via the axum path extractor, whereas we want it to return a `shuttle_gateway::Error`. This means if a user with an older `cargo-shuttle` calls the gateway after this change, it will be parsed to 400 status and "this request is invalid". It will not have the message regarding that this is an invalid project name and the rules for project names.

Ideally we would figure out a way to have this failure return a gateway `Error`, perhaps by creating a custom path extractor, or by not parsing the project name in deserialize, but rather a custom extractor for ProjectName. I'll look into it.

Reply:
From what I remember, project name were not even validated in the extractor. Not sure though. I'm in favor of explicit validation in all places it is needed (not on deser).

Reply:
I tested it and it did error in deserialize, but it didn't return an error that the CLI could parse. And making it return a gateway::Error seemed quite complicated. I am also in favor of moving it out of deserialize, and rather using a conventional newtype pattern where a ProjectName can only be created by parsing a valid project name. Perhaps a custom extractor in the API.

Reply:
Removing the parse from deserialization sounds reasonable. I don't have an understanding of the gateway, so I think this up to you.

Reply:
> user with an older cargo-shuttle

The older ones do client side validation, so is this an issue?